core/src/kura.rs

```diff
@@ -536,9 +415,48 @@ pub struct BlockIndex {
     pub length: u64,
 }
 
-/// Operations available both on locked and unlocked store,
-/// i.e. only read operations.
-impl<L: Lock> BlockStore<L> {
+impl BlockStore {
+    /// Create a new read-only block store in `path`.
+    pub fn new(path: &Path, already_locked: bool) -> Self {
+        if !already_locked {
+            let path = path.join(LOCK_FILE_NAME);
+            if let Err(e) = fs::File::options()
+                .read(true)
+                .write(true)
+                .create_new(true)
+                .open(path.clone())
+            {
+                match e.kind() {
+                    std::io::ErrorKind::AlreadyExists => Err(Error::Locked(path)),
+                    std::io::ErrorKind::NotFound => {
+                        match std::fs::create_dir_all(&path)
+                            .map_err(|e| Error::MkDir(e, path.clone()))
+                        {
+                            Err(e) => Err(e),
+                            Ok(_) => {
+                                if let Err(e) = fs::File::options()
+                                    .read(true)
+                                    .write(true)
+                                    .create_new(true)
+                                    .open(path.clone())
+                                {
+                                    Err(Error::IO(e, path))
+                                } else {
+                                    Ok(())
+                                }
+                            }
+                        }
+                    }
+                    _ => Err(Error::IO(e, path)),
+                }
+                .expect("Kura must be able to lock the blockstore");
+            }
+        }
+        BlockStore {
+            path_to_blockchain: path.to_path_buf(),
+        }
```

Comment:
```suggestion
     else {
        BlockStore {
            path_to_blockchain: path.to_path_buf(),
        }
     }
```

The control flow is a bit confusing with this level of nesting.

Reply:
I reordered to make it slightly easier on the eye.